From 3d601ed9e8e4f4d1d037a0f3c0525b9965124e17 Mon Sep 17 00:00:00 2001
From: Ariel Elior <ariele@broadcom.com>
Date: Wed, 17 Apr 2013 22:49:06 +0000
Subject: [PATCH 614/858] bnx2x: Fix status blocks configuration

This fixes 2 issues regarding bnx2x's status blocks:

   1. ethtool -c caused corruption of status blocks in FW RAM.

   2. when using multi-CoS, the configuration of the timeout values of
      status blocks is incorrect, harming the coalescing of interrupts
      for such CoSs.

Signed-off-by: Ariel Elior <ariele@broadcom.com>
Signed-off-by: Yuval Mintz <yuvalmin@broadcom.com>
Signed-off-by: Eilon Greenstein <eilong@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 0c14e5ced264620284bd96f888614768c9bd0976)

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
index 7060847..3ce49a1 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
@@ -2564,6 +2564,9 @@ int bnx2x_nic_load(struct bnx2x *bp, int load_mode)
    }
   }
 
+  /* initialize FW coalescing state machines in RAM */
+  bnx2x_update_coalesce(bp);
+
   /* setup the leading queue */
   rc = bnx2x_setup_leading(bp);
   if (rc) {
@@ -4530,11 +4533,11 @@ static void storm_memset_hc_disable(struct bnx2x *bp, u8 port,
  u32 enable_flag = disable ? 0 : (1 << HC_INDEX_DATA_HC_ENABLED_SHIFT);
  u32 addr = BAR_CSTRORM_INTMEM +
      CSTORM_STATUS_BLOCK_DATA_FLAGS_OFFSET(fw_sb_id, sb_index);
- u16 flags = REG_RD16(bp, addr);
+ u8 flags = REG_RD8(bp, addr);
  /* clear and set */
  flags &= ~HC_INDEX_DATA_HC_ENABLED;
  flags |= enable_flag;
- REG_WR16(bp, addr, flags);
+ REG_WR8(bp, addr, flags);
  DP(NETIF_MSG_IFUP,
     "port %x fw_sb_id %d sb_index %d disable %d\n",
     port, fw_sb_id, sb_index, disable);
-- 
1.8.3.1